openmmtools/alchemy.py

```diff
@@ -12,413 +12,905 @@
+    >>> print(alchemical_state.lambda_sterics)
+    None
+    >>> alchemical_state.apply_to_system(alanine_alchemical_system)
+    Traceback (most recent call last):
+    ...
+    AlchemicalStateError: The system parameter lambda_sterics is not defined in this state.
+
+    Create an AlchemicalState that matches the parameters defined in
+    the System.
+
+    >>> alchemical_state = AlchemicalState.from_system(alanine_alchemical_system)
+    >>> alchemical_state.lambda_sterics
+    1.0
+    >>> alchemical_state.lambda_electrostatics
+    1.0
+    >>> print(alchemical_state.lambda_angles)
+    None
+
+    AlchemicalState implement the IComposableState interface, so it can be
+    used with CompoundThermodynamicState. All the alchemical parameters are
+    accessible through the compound state.
+
+    >>> from simtk import openmm, unit
+    >>> thermodynamic_state = states.ThermodynamicState(system=alanine_alchemical_system,
+    ...                                                 temperature=300*unit.kelvin)
+    >>> compound_state = states.CompoundThermodynamicState(thermodynamic_state=thermodynamic_state,
+    ...                                                    composable_states=[alchemical_state])
+    >>> compound_state.lambda_sterics
+    1.0
+
+    You can control the parameters in the OpenMM Context in this state by
+    setting the state attributes.
+
+    >>> compound_state.lambda_sterics = 0.5
+    >>> integrator = openmm.VerletIntegrator(1.0*unit.femtosecond)
+    >>> context = compound_state.create_context(integrator)
+    >>> context.getParameter('lambda_sterics')
+    0.5
+    >>> compound_state.lambda_sterics = 1.0
+    >>> compound_state.apply_to_context(context)
+    >>> context.getParameter('lambda_sterics')
+    1.0
+
+    You can express the alchemical parameters as a mathematical expression
+    involving alchemical variables. Here is an example for a two-stage function.
+
+    >>> compound_state.set_alchemical_variable('lambda', 1.0)
+    >>> compound_state.lambda_sterics = AlchemicalFunction('step_hm(lambda - 0.5) + 2*lambda * step_hm(0.5 - lambda)')
+    >>> compound_state.lambda_electrostatics = AlchemicalFunction('2*(lambda - 0.5) * step(lambda - 0.5)')
+    >>> for l in [0.0, 0.25, 0.5, 0.75, 1.0]:
+    ...     compound_state.set_alchemical_variable('lambda', l)
+    ...     print(compound_state.lambda_sterics)
+    0.0
+    0.5
+    1.0
+    1.0
+    1.0
 
-    Periodic water box.
 
-    >>> waterbox = testsystems.WaterBox()
-    >>> print(_is_periodic(waterbox.system))
-    True
+    """
 
-    Non-periodic Lennard-Jones cluster.
+    # -------------------------------------------------------------------------
+    # Constructors
+    # -------------------------------------------------------------------------
 
-    >>> cluster = testsystems.LennardJonesCluster()
-    >>> print(_is_periodic(cluster.system))
-    False
+    def __init__(self, **kwargs):
+        self._alchemical_variables = {}
```

Comment:
Alchemical variables are those extra parameters that can be used inside an expression in `AlchemicalFunction`.